crates/storage/nippy-jar/src/lib.rs

```diff
@@ -33,36 +33,68 @@ const NIPPY_JAR_VERSION: usize = 1;
 /// A [`Row`] is a list of its selected column values.
 type Row = Vec<Vec<u8>>;
 
+/// Placeholder type for when a jar has no user header.
+pub type NoJarHeader = bool;
+
+/// `NippyJar` is a specialized storage format designed for immutable data.
+///
+/// Data is organized into a columnar format, enabling column-based compression. Data retrieval
+/// entails consulting an offset list and fetching the data from file via `mmap`.
+///
+/// PHF & Filters:
+/// For data membership verification, the `filter` field can be configured with algorithms like
+/// Bloom or Cuckoo filters. While these filters enable rapid membership checks, it's important to
+/// note that **they may yield false positives but not false negatives**. Therefore, they serve as
+/// preliminary checks (eg. in `by_hash` queries) and should be followed by data verification on
+/// retrieval.
+///
+/// The `phf` (Perfect Hashing Function) and `offsets_index` fields facilitate the data retrieval
+/// process in for example `by_hash` queries. Specifically, the PHF converts a query, such as a
+/// block hash, into a unique integer. This integer is then used as an index in `offsets_index`,
+/// which maps to the actual data location in the `offsets` list. Similar to the `filter`, the PHF
+/// may also produce false positives but not false negatives, necessitating subsequent data
+/// verification.
+///
+/// Note: that the key (eg. BlockHash) passed to a filter and phf does not need to actually be
+/// stored.
+///
+/// Ultimately, the `freeze` function yields two files: a data file containing both the data and its
+/// configuration, and an index file that houses the offsets and offsets_index.
 #[derive(Debug, Serialize, Deserialize)]
 #[cfg_attr(test, derive(PartialEq))]
-pub struct NippyJar {
-    /// Version
+pub struct NippyJar<H> {
```

Comment:
this leads to weird API, see `NoJarHeader` helper type

can we use a default value = `()` here?

also, how does this generic affect `NippyJar::load`?

Reply:
the user needs to ensure that the type is the same as in `new`